third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp

```diff
@@ -146,17 +147,88 @@ Operation *createIglpOpt(PatternRewriter &rewriter, Location loc, int value) {
   return rewriter.create<ROCDL::IglpOpt>(loc, iglpValue);
 }
 
+// The following structs represent in-source database regarding a target
+// machine. It provides instructions execution and issue cycles needed for
+// scheduling.
+struct MachineDescr {
+  virtual ~MachineDescr() = default;
+  virtual uint32_t getDsReadIssueCycle(uint32_t instrWidth) = 0;
+  virtual FailureOr<uint32_t> getMmaExecCycle(llvm::ArrayRef<int64_t> dims) = 0;
+  virtual uint32_t getMmaIssueCycle() = 0;
+  virtual uint32_t getNumLdsDataPaths() = 0;
+  static std::unique_ptr<MachineDescr> get(std::string arch);
+};
+
+template <typename Derived> struct MachineDescrImpl : MachineDescr {
+  uint32_t getDsReadIssueCycle(uint32_t instrWidth) final {
+    return instrWidth == 16 ? 8 : 4;
+  }
+
+  FailureOr<uint32_t> getMmaExecCycle(llvm::ArrayRef<int64_t> dims) final {
+    if (dims.size() != 3)
+      return failure();
+    auto it =
+        Derived::mmaTable.find(std::make_tuple(dims[0], dims[1], dims[2]));
+    if (it != Derived::mmaTable.end())
+      return it->second;
+    return failure();
+  }
+
+  uint32_t getMmaIssueCycle() final { return Derived::mmaIssueCycle; };
+  uint32_t getNumLdsDataPaths() final { return Derived::numLdsDataPaths; }
+
+  using MmaTable =
+      llvm::DenseMap<std::tuple<int64_t, int64_t, int64_t>, uint32_t>;
+};
+
+struct MI200Kind : public MachineDescrImpl<MI200Kind> {
+  static const inline MmaTable mmaTable{{{32, 32, 8}, 64}, {{16, 16, 16}, 32}};
+  static const inline uint32_t mmaIssueCycle{4};
+  static const inline uint32_t numLdsDataPaths{2};
+};
+
+struct MI300Kind : public MachineDescrImpl<MI300Kind> {
+  static const inline MmaTable mmaTable{{{32, 32, 8}, 32}, {{16, 16, 16}, 16}};
+  static const inline uint32_t mmaIssueCycle{4};
+  static const inline uint32_t numLdsDataPaths{2};
+};
+
+std::unique_ptr<MachineDescr> MachineDescr::get(std::string arch) {
+  llvm::AMDGPU::GPUKind archKind = llvm::AMDGPU::parseArchAMDGCN(arch);
+
+  switch (archKind) {
+  case llvm::AMDGPU::GPUKind::GK_GFX940: {
+    [[fallthrough]];
+  };
+  case llvm::AMDGPU::GPUKind::GK_GFX941: {
+    [[fallthrough]];
+  };
+  case llvm::AMDGPU::GPUKind::GK_GFX942: {
+    return std::make_unique<MachineDescrImpl<MI300Kind>>();
+  }
+  case llvm::AMDGPU::GPUKind::GK_GFX90A: {
+    return std::make_unique<MachineDescrImpl<MI200Kind>>();
+  }
+  default: {
+    return nullptr;
+  }
+  }
+  return nullptr;
+}
+
 struct InstructionSchedHintsRewriter
     : public OpRewritePattern<triton::amdgpu::InstructionSchedHint> {
 
-  InstructionSchedHintsRewriter(MLIRContext *ctx, int32_t numStages,
-                                std::string variant)
+  InstructionSchedHintsRewriter(MLIRContext *ctx, std::string arch,
```

Comment:
Nit: use `StringRef` to avoid unnecessary copies. Similarly for following cases.

Reply:
Done